nairuby/arc_platform
# Make learning materials page CRUD using the Rails Admin gem

# Hello there :wave: 

This pull request works on part of the requirements in [this issue](#78 ). I tried to stick as much as possible to the [provided design guidelines](https://still-snowflake-8822.animaapp.io/materials).


The specific tasks implemented in this pull request include:
1. :one:  Setup of the [Rails Admin gem](https://github.com/railsadminteam/rails_admin) to the project which provides a dashboard. The dashboard allows a community admin to create, edit, and delete records. and Learning materials (and other resources as well in the project.). The following sub-sections show the process of creating, editing, and deleting a learning material.
     a. Creating a learning material record from the Rails Admin dashboard
      
[Screencast from 17-04-2024 08:36:05 ALASIRI.webm](https://github.com/nairuby/arc_platform/assets/86318284/957d066b-d8c8-427e-8f02-5698e1342bd5)

     b. Updating a learning material record from the Rails Admin dashboard
[Screencast from 17-04-2024 08:37:17 ALASIRI.webm](https://github.com/nairuby/arc_platform/assets/86318284/1afc4d3a-281d-49c2-9bb7-aea9b1d394e5)

     c. Deleting a learning material record from the Rails Admin dashboard
[Screencast from 17-04-2024 08:38:14 ALASIRI.webm](https://github.com/nairuby/arc_platform/assets/86318284/e62a850c-01fb-43f7-9f87-75914933ab88)

### The screencast below shows what the new learning materials page looks like:

[Screencast from 17-04-2024 08:41:19 ALASIRI.webm](https://github.com/nairuby/arc_platform/assets/86318284/4d7323c9-a204-41aa-9d21-ff5fc5b14307)


2.  :two:  Authorisation has been implemented using the cancancan gem which makes sure that only a community admin with the role "organization_admin" can access the dashboard and make any necessary changes.

3. :three: Updated the routes including the /admin and /learning_materials routes to allow access to the new features

4. :four: Created a ```learning_materials/index.html.erb``` view which holds the markup for the new learning page. The view is designed using TailwindCSS and is mobile responsive as well.

## Other changes made that were not part of the core requirements:
1. :one: In the ```stylesheets/mailgun_mails.css``` file, I noticed that there are repetitions of some color codes. I therefore made some CSS variables to reduce these many repetitions.

2. :two: Added Rspec and Capybara for some integration tests.

## How to test the changes made:
1. :one: To best the new learning materials page, navigate to this page ```http://127.0.0.1:3000/learning_materials```vor optionally click on any link that has "Learning Materials```.

2. :two: To navigate to the admin dashboard:
    a. Ensure that you have an admin account. You can create an account and edit it in the dashboard under **"users"**. Make sure that the account role is ```organization_admin```. To do this you might need to first disable lines 45-49 in the ```config/initializers/rails_admin.rb``` file. Then un comment it afterward to test other bits of the changes made.
   b. Once you have an account that has the organization admin role, you can now navigate to ```http://127.0.0.1:3000/admin``` after you have successfully authenticated yourself.

3. To run the tests run ```bundle exec rspec spec/features/learning_materials_spec.rb```.

## Features not yet implemented in this pull request:
Having a look at the provided [design guidelines](https://still-snowflake-8822.animaapp.io/materials), the feature for "searching" to filter the available learning materials has not yet been implemented. This could be implemented as a feature improvement of the existing code.

## Slight bug known which does not affect the working of the feature implemented:
When you access the learning materials resource, you will notice two "thumbnails". One as an input box and one as an image upload. This is caused because:
1. I have the thumbnails column as a string (which shows the input form for text in the rails admin dashboard).
2. In the learning materials model, I have the ```has_one_attached``` association which creates an Active Record attachment which creates the other thumbnail.

**Therefore when creating the learning materials from the admin dashboard you can safely ignore the one that shows as an image upload. I would appreciate suggestions to make the thumbnail show as one and not two in the admin dashboard.**


### Thank you for taking the time to check out my PR 🙏

![](https://media.giphy.com/media/v1.Y2lkPTc5MGI3NjExYnJ1ejQ5eGdsMGtoZG1pMnNiMXN4N2l4MzJhejd3MjRoOWsxOG4zZSZlcD12MV9pbnRlcm5hbF9naWZfYnlfaWQmY3Q9Zw/ZfK4cXKJTTay1Ava29/giphy.gif)
